### DISCUSSION TITLE
Uncaught Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate.

### DISCUSSION THREAD

**Discussion by madanbajagai3** [clauderic/dnd-kit#1244] [Q&A]
DragAndDropBoard.js
`import {
  DndContext,
  DragOverlay,
  PointerSensor,
  useSensor,
  useSensors,
} from "@dnd-kit/core";
import React, { useMemo, useState } from "react";
import { SortableContext, arrayMove } from "@dnd-kit/sortable";
import ColumnContainer from "./ColumnContainer";
import { createPortal } from "react-dom";
import TaskCard from "./TaskCard";
import "./DragAndDropBoard.css";

const DragAndDropBoard = () => {
  function generateId() {
    return Math.floor(Math.random() * 10001);
  }

  const defaultCols = [
    {
      id: "todo",
      title: "Todo",
    },
    {
      id: "doing",
      title: "Work in progress",
    },
    {
      id: "done",
      title: "Done",
    },
  ];

  const defaultTasks = [
    {
      id: "1",
      columnId: "todo",
      content: "List admin APIs for dashboard",
    },
    {
      id: "2",
      columnId: "todo",
      content:
        "Develop user registration functionality with OTP delivered on SMS after email confirmation and phone number confirmation",
    },
    {
      id: "3",
      columnId: "doing",
      content: "Conduct security testing",
    },
    {
      id: "4",
      columnId: "doing",
      content: "Analyze competitors",
    },
    {
      id: "5",
      columnId: "done",
      content: "Create UI kit documentation",
    },
    {
      id: "6",
      columnId: "done",
      content: "Dev meeting",
    },
    {
      id: "7",
      columnId: "done",
      content: "Deliver dashboard prototype",
    },
    {
      id: "8",
      columnId: "todo",
      content: "Optimize application performance",
    },
    {
      id: "9",
      columnId: "todo",
      content: "Implement data validation",
    },
    {
      id: "10",
      columnId: "todo",
      content: "Design database schema",
    },
    {
      id: "11",
      columnId: "todo",
      content: "Integrate SSL web certificates into workflow",
    },
    {
      id: "12",
      columnId: "doing",
      content: "Implement error logging and monitoring",
    },
    {
      id: "13",
      columnId: "doing",
      content: "Design and implement responsive UI",
    },
  ];

  const [columns, setColumns] = useState(defaultCols);

  const columnsId = useMemo(() => columns.map((col) => col.id), [columns]);

  const [tasks, setTasks] = useState(defaultTasks);

  const [activeColumn, setActiveColumn] = useState(null);

  const [activeTask, setActiveTask] = useState(null);

  function createNewColumn() {
    const columnToAdd = {
      id: generateId(),
      title: `Column ${columns.length + 1}`,
    };

    setColumns([...columns, columnToAdd]);
  }
  function updateColumn(id, title) {
    const newColumns = columns.map((col) => {
      if (col.id !== id) return col;
      return { ...col, title };
    });

    setColumns(newColumns);
  }
  function deleteColumn(id) {
    const filteredColumns = columns.filter((col) => col.id !== id);
    setColumns(filteredColumns);

    const newTasks = tasks.filter((t) => t.columnId !== id);
    setTasks(newTasks);
  }
  function createTask(columnId) {
    const newTask = {
      id: generateId(),
      columnId,
      content: `Task ${tasks.length + 1}`,
    };

    setTasks([...tasks, newTask]);
  }
  function deleteTask(id) {
    const newTasks = tasks.filter((task) => task.id !== id);
    setTasks(newTasks);
  }
  function updateTask(id, content) {
    const newTasks = tasks.map((task) => {
      if (task.id !== id) return task;
      return { ...task, content };
    });

    setTasks(newTasks);
  }

  const sensors = useSensors(
    useSensor(PointerSensor, {
      activationConstraint: {
        distance: 10,
      },
    })
  );

  function onDragStart(event) {
    if (event.active.data.current?.type === "Column") {
      setActiveColumn(event.active.data.current.column);
      return;
    }

    if (event.active.data.current?.type === "Task") {
      setActiveTask(event.active.data.current.task);
      return;
    }
  }

  function onDragEnd(event) {
    setActiveColumn(null);
    setActiveTask(null);

    const { active, over } = event;
    if (!over) return;

    const activeId = active.id;
    const overId = over.id;

    if (activeId === overId) return;

    const isActiveAColumn = active.data.current?.type === "Column";
    if (!isActiveAColumn) return;

    setColumns((columns) => {
      const activeColumnIndex = columns.findIndex((col) => col.id === activeId);

      const overColumnIndex = columns.findIndex((col) => col.id === overId);

      return arrayMove(columns, activeColumnIndex, overColumnIndex);
    });
  }

  function onDragOver(event) {
    const { active, over } = event;
    if (!over) return;

    const activeId = active.id;
    const overId = over.id;

    if (activeId === overId) return;

    const isActiveATask = active.data.current?.type === "Task";
    const isOverATask = over.data.current?.type === "Task";

    if (!isActiveATask) return;

    // Im dropping a Task over another Task
    if (isActiveATask && isOverATask) {
      setTasks((tasks) => {
        const activeIndex = tasks.findIndex((t) => t.id === activeId);
        const overIndex = tasks.findIndex((t) => t.id === overId);

        if (tasks[activeIndex].columnId != tasks[overIndex].columnId) {
          // Fix introduced after video recording
          tasks[activeIndex].columnId = tasks[overIndex].columnId;
          return arrayMove(tasks, activeIndex, overIndex - 1);
        }

        return arrayMove(tasks, activeIndex, overIndex);
      });
    }

    const isOverAColumn = over.data.current?.type === "Column";

    // Im dropping a Task over a column
    if (isActiveATask && isOverAColumn) {
      setTasks((tasks) => {
        const activeIndex = tasks.findIndex((t) => t.id === activeId);

        tasks[activeIndex].columnId = overId;
        return arrayMove(tasks, activeIndex, activeIndex);
      });
    }
  }

  return (
    <div>
      <div className="board">
        <DndContext
          sensors={sensors}
          onDragStart={onDragStart}
          onDragEnd={onDragEnd}
          onDragOver={onDragOver}
        >
          <SortableContext items={columnsId}>
            {columns.map((col) => (
              <ColumnContainer
                key={col.id}
                column={col}
                deleteColumn={deleteColumn}
                updateColumn={updateColumn}
                createTask={createTask}
                deleteTask={deleteTask}
                updateTask={updateTask}
                tasks={tasks.filter((task) => task.columnId === col.id)}
              />
            ))}
            <button className="card_btn" onClick={createNewColumn}>
              Add Column
            </button>
          </SortableContext>
          {createPortal(
            <DragOverlay>
              {activeColumn && (
                <ColumnContainer
                  column={activeColumn}
                  deleteColumn={deleteColumn}
                  updateColumn={updateColumn}
                  createTask={createTask}
                  deleteTask={deleteTask}
                  updateTask={updateTask}
                  tasks={tasks.filter(
                    (task) => task.columnId === activeColumn.id
                  )}
                />
              )}
              {activeTask && (
                <TaskCard
                  task={activeTask}
                  deleteTask={deleteTask}
                  updateTask={updateTask}
                />
              )}
            </DragOverlay>,
            document.body
          )}
        </DndContext>
      </div>
    </div>
  );
};

export default DragAndDropBoard;
`

ColumnContainer.js
`import { SortableContext, useSortable } from "@dnd-kit/sortable";
import React, { useMemo, useState } from "react";
import TaskCard from "./TaskCard";
import "./DragAndDropBoard.css";
import { BsThreeDots } from "react-icons/bs";

import { CSS } from "@dnd-kit/utilities";
import { AiFillPlusCircle } from "react-icons/ai";
import { MdDelete } from "react-icons/md";

const ColumnContainer = ({
  column,
  deleteColumn,
  updateColumn,
  createTask,
  tasks,
  deleteTask,
  updateTask,
}) => {
  const [editMode, setEditMode] = useState(false);

  const tasksIds = useMemo(() => {
    return tasks.map((task) => task.id);
  }, [tasks]);

  const {
    setNodeRef,
    attributes,
    listeners,
    transform,
    transition,
    isDragging,
  } = useSortable({
    id: column.id,
    data: {
      type: "Column",
      column,
    },
    disabled: editMode,
  });

  const style = {
    transition,
    transform: CSS.Transform.toString(transform),
  };

  if (isDragging) {
    return (
      <div ref={setNodeRef} style={style} className="card_IsDragging"></div>
    );
  }

  return (
    <div ref={setNodeRef} style={style} className="taskColumn">
      {/* Column title */}
      <div
        {...attributes}
        {...listeners}
        onClick={() => {
          setEditMode(true);
        }}
        className="colTitle"
      >
        <div className="colTitle_div">
          {!editMode && `${column.title} (${tasks.length})`}
          {editMode && (
            <input
              className="bg-black focus:border-rose-500 border rounded outline-none px-2"
              value={column.title}
              onChange={(e) => updateColumn(column.id, e.target.value)}
              autoFocus
              onBlur={() => {
                setEditMode(false);
              }}
              onKeyDown={(e) => {
                if (e.key !== "Enter") return;
                setEditMode(false);
              }}
            />
          )}
        </div>
        <span>
          <BsThreeDots />
        </span>
      </div>

      <div className="tasks_container ">
        {tasks?.map((task) => (
          <TaskCard key={task.id} task={task} />
        ))}
      </div>
      <h5 className="text-center font-weight-bold mt-3">Total: $5000</h5>
    </div>
  );
};

export default ColumnContainer;
`

TaskCard.js
`import { useSortable } from "@dnd-kit/sortable";
import React, { useState } from "react";
import "./DragAndDropBoard.css";
import { AiFillPlusCircle } from "react-icons/ai";
import avatar1 from "../../assets/crm/pipeline/avatar.png";
import avatar2 from "../../assets/crm/pipeline/avatar2.png";

import { CSS } from "@dnd-kit/utilities";
import Modal from "../Modal/Modal";
import DndModal from "./DndModal";

const TaskCard = ({ task, deleteTask, updateTask, title }) => {
  const [mouseIsOver, setMouseIsOver] = useState(false);
  const [editMode, setEditMode] = useState(false);
  const [showModal, setShowModal] = useState(false);

  const handleModal = () => {
    setShowModal((prev) => !prev);
  };

  const {
    setNodeRef,
    attributes,
    listeners,
    transform,
    transition,
    isDragging,
  } = useSortable({
    id: task.id,
    data: {
      type: "Task",
      task,
    },
    disabled: showModal,
  });

  const style = {
    transition,
    transform: CSS.Transform.toString(transform),
  };

  if (isDragging) {
    return <div ref={setNodeRef} style={style} className=" item_IsDragging" />;
  }

  if (showModal) {
    return (
      <div
        className="textareaDiv"
        ref={setNodeRef}
        style={style}
        {...attributes}
        {...listeners}
      >
        <Modal
          showModal={showModal}
          setShowModal={setShowModal}
          size={"modal-xl"}
          header={title}
        >
          <DndModal content={task.content} />
        </Modal>
      </div>
    );
  }

  return (
    <>
      <div
        className="taskCard"
        ref={setNodeRef}
        style={style}
        {...attributes}
        {...listeners}
        onDoubleClick={handleModal}
      >
        <div className="d-flex w-100 ">
          <div className="task_verticalLine"></div>
          <div className="w-100 py-1 ">
            <div className="d-flex w-100 ">
              <p className="w-100 font-weight-bold">Transfer to Sikhar</p>
              <div onClick={handleModal}>
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  width="20"
                  height="20"
                  viewBox="0 0 20 20"
                  style={{ height: "25px", width: "25px" }}
                  fill="white"
                >
                  <path
                    d="M4.66797 4.6665H4.0013C3.64768 4.6665 3.30854 4.80698 3.05849 5.05703C2.80844 5.30708 2.66797 5.64622 2.66797 5.99984V11.9998C2.66797 12.3535 2.80844 12.6926 3.05849 12.9426C3.30854 13.1927 3.64768 13.3332 4.0013 13.3332H10.0013C10.3549 13.3332 10.6941 13.1927 10.9441 12.9426C11.1942 12.6926 11.3346 12.3535 11.3346 11.9998V11.3332"
                    stroke="#393B40"
                    strokeWidth="2"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                  />
                  <path
                    d="M10.6667 3.3334L12.6667 5.3334M13.59 4.39007C13.8526 4.12751 14.0001 3.77139 14.0001 3.40007C14.0001 3.02875 13.8526 2.67264 13.59 2.41007C13.3274 2.14751 12.9713 2 12.6 2C12.2287 2 11.8726 2.14751 11.61 2.41007L6 8.00007V10.0001H8L13.59 4.39007Z"
                    stroke="#393B40"
                    strokeWidth="2"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                  />
                </svg>
              </div>
            </div>
            <div className="d-flex w-100 justify-content-between font-size-15">
              <p>Soori Solutions</p>
              <p>Amount $1000</p>
            </div>
            <p style={{ color: "#2A66FF" }} className="w-100 ">
              Banking Sector
            </p>
            <hr className="mx-0 my-1" />
            <div className="d-flex">
              <p className="w-100 font-size-15">Aug 16 - Aug 20</p>
              <div className="avatars d-flex">
                <img src={avatar1} alt="" />
                <img src={avatar2} alt="" />
              </div>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default TaskCard;
![Screenshot from 2023-10-03 14-13-52](https://github.com/clauderic/dnd-kit/assets/143921330/56fc327d-b6ca-4471-a334-2a7673c6769d)
![Screenshot from 2023-10-03 14-13-40](https://github.com/clauderic/dnd-kit/assets/143921330/4d034039-aace-44ee-ba82-c0aeed84ca8b)
![Screenshot from 2023-10-03 14-13-24](https://github.com/clauderic/dnd-kit/assets/143921330/07413d71-fff7-455c-8836-159031d2a4ed)
`

**Comment by bycolour2** [comment]
Try to rewatch video and dig into source code on git hub repo, as i know author missed few things in video and fixed it in github repo
